From 56ab24b94fd5f090fa4888249c3f5b051a749626 Mon Sep 17 00:00:00 2001
From: N Balachandran <nbalacha@redhat.com>
Date: Mon, 11 Jan 2016 12:13:32 +0530
Subject: [PATCH 238/245] cluster/dht : Rebalance process crashes due to double fd_unref

The dst_fd was being unrefed twice in case the call
to __dht_rebalance_create_dst_file failed.

Change-Id: Iba8b0d26a0e916a70a0ce27e31c8e3120a2c12cb
BUG: 1296134
upstream: http://review.gluster.org/#/c/13193/
Signed-off-by: N Balachandran <nbalacha@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/65199
Reviewed-by: Dan Lambright <dlambrig@redhat.com>
Tested-by: Dan Lambright <dlambrig@redhat.com>
---
 xlators/cluster/dht/src/dht-rebalance.c |   17 +++++++++++------
 1 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/xlators/cluster/dht/src/dht-rebalance.c b/xlators/cluster/dht/src/dht-rebalance.c
index 20e9cd2..c095850 100644
--- a/xlators/cluster/dht/src/dht-rebalance.c
+++ b/xlators/cluster/dht/src/dht-rebalance.c
@@ -519,7 +519,7 @@ __dht_rebalance_create_dst_file (xlator_t *to, xlator_t *from, loc_t *loc, struc
                                 DHT_MSG_GFID_MISMATCH,
                                 "file %s exists in %s with different gfid",
                                 loc->path, to->name);
-                        fd_unref (fd);
+                        ret = -1;
                         goto out;
                 }
         }
@@ -563,8 +563,7 @@ __dht_rebalance_create_dst_file (xlator_t *to, xlator_t *from, loc_t *loc, struc
         }
 
         fd_bind (fd);
-        if (dst_fd)
-                *dst_fd = fd;
+
         /*Reason of doing lookup after create again:
          *In the create, there is some time-gap between opening fd at the
          *server (posix_layer) and binding it in server (incrementing fd count),
@@ -575,7 +574,6 @@ __dht_rebalance_create_dst_file (xlator_t *to, xlator_t *from, loc_t *loc, struc
          *at  the backend.
          */
 
-
         ret = syncop_lookup (to, loc, &check_stbuf, NULL, NULL, NULL);
         if (!ret) {
 
@@ -586,7 +584,6 @@ __dht_rebalance_create_dst_file (xlator_t *to, xlator_t *from, loc_t *loc, struc
                                 "found in lookup after create",
                                 loc->path, to->name);
                         ret = -1;
-                        fd_unref (fd);
                         goto out;
                 }
 
@@ -597,7 +594,6 @@ __dht_rebalance_create_dst_file (xlator_t *to, xlator_t *from, loc_t *loc, struc
                         DHT_MSG_MIGRATE_FILE_FAILED, "%s: file does not exists"
                         "on %s (%s)", loc->path, to->name, strerror (-ret));
                 ret = -1;
-                fd_unref (fd);
                 goto out;
         }
 
@@ -627,7 +623,15 @@ __dht_rebalance_create_dst_file (xlator_t *to, xlator_t *from, loc_t *loc, struc
         /* success */
         ret = 0;
 
+        if (dst_fd)
+                *dst_fd = fd;
+
 out:
+        if (ret) {
+                if (fd) {
+                        fd_unref (fd);
+                }
+        }
         if (dict)
                 dict_unref (dict);
 
@@ -898,6 +902,7 @@ __dht_rebalance_open_src_file (xlator_t *from, xlator_t *to, loc_t *loc,
         }
 
         fd_bind (fd);
+
         if (src_fd)
                 *src_fd = fd;
 
-- 
1.7.1

